[DOCS] Fix data stream docs (#59818)
elastic/elasticsearch: docs/reference/data-streams/data-streams.asciidoc

[discrete]
[[backing-indices]]
== Backing indices
A data stream consists of one or more _backing indices_. Backing indices are
<<index-hidden,hidden>>, auto-generated indices used to store a stream's
documents.

image::images/data-streams/data-streams-diagram.svg[align="center"]

To create backing indices, each data stream requires a matching
<<indices-templates,index template>>. This template acts as a blueprint for the
stream's backing indices. It contains:

* The mappings and settings applied to each backing index when it's created.

* A name or wildcard (`*`) pattern that matches the data stream's name.

* A `data_stream` object with an empty body (`{ }`). This object indicates the
template is used for data streams.

A `@timestamp` field must be included in every document indexed to the data
stream. This field can be mapped as a <<date,`date`>> or
<<date_nanos,`date_nanos`>> field data type in the stream's matching index
template. If no mapping is specified in the template, the `date` field data type
with default options is used.

The same index template can be used to create multiple data streams.

[discrete]
[[data-streams-generation]]

> == Generation

Reviewer: Maybe Data stream generations? On it's own, it brings to mind generating data streams.

Each data stream tracks its _generation_: a six-digit, zero-padded integer
that acts as a cumulative count of the data stream's backing indices. This count
includes any deleted indices for the stream. The generation is incremented
whenever a new backing index is added to the stream.

When a backing index is created, the index is named using the following
convention:

[source,text]
----
.ds-<data-stream>-<generation>
----

For example, the `web_server_logs` data stream has a generation of `34`. The
most recently created backing index for this data stream is named
`.ds-web_server_logs-000034`.

Because the generation increments with each new backing index, backing indices
with a higher generation contain more recent data. Backing indices with a lower
generation contain older data.

> A backing index's name can change after its creation due to a
> <<indices-shrink-index,shrink>>, <<snapshots-restore-snapshot,restore>>, or
> other operations.

Reviewer: Maybe:

```suggestion
Performing operations such as a <<indices-shrink-index,shrink>> or <<snapshots-restore-snapshot,restore>> can change the name of a backing index.
```
It feels like maybe there's another point here--like changing the name of a backing index sort of "detaches" it from the data stream.

[discrete]
[[data-stream-write-index]]
== Write index

When a read request is sent to a data stream, it routes the request to all its
backing indices. For example, a search request sent to a data stream would query
all its backing indices.

image::images/data-streams/data-streams-search-request.svg[align="center"]

> However, the most recently created backing index is the data stream’s only

Reviewer: I would lead with this part and move the part about federating the search across all of the backing indices afterward. Otherwise, it's a bit jarring to start reading about read requests in a section about write requests.

* <<indices-close,Close>>
* <<indices-delete-index,Delete>>
* <<freeze-index-api,Freeze>>
* <<indices-shrink-index,Shrink>>
* <<indices-split-index,Split>>

[discrete]
[[data-streams-rollover]]
== Rollover

When a data stream is created, one backing index is automatically created.
Because this single index is also the most recently created backing index, it
acts as the stream's write index.

A <<indices-rollover-index,rollover>> creates a new backing index for a data
stream. This new backing index becomes the stream's write index, replacing
the current one, and increments the stream's generation.

In most cases, we recommend using <<index-lifecycle-management,{ilm}
({ilm-init})>> to automate rollovers for data streams. This lets you
automatically roll over the current write index when it meets specified
criteria, such as a maximum age or size.

However, you can also use the <<indices-rollover-index,rollover API>> to
manually perform a rollover. See <<manually-roll-over-a-data-stream>>.

[discrete]
[[data-streams-append-only]]
== Append-only

For most time-series use cases, existing data is rarely, if ever, updated.
Because of this, data streams are designed to be append-only.

> You can send <<add-documents-to-a-data-stream,indexing requests for new

Reviewer 2: Would the more detailed information on supported APIs fit better in the 'Use a data stream' docs, to keep this shorter and high level?

documents>> directly to a data stream. However, you cannot send the following
requests for existing documents directly to a data stream:

* An <<docs-index_,index API>> request with an
  <<docs-index-api-op_type,`op_type`>> of `index`. The `op_type` parameter
  defaults to `index` for existing documents.

* A <<docs-bulk,bulk API>> request using the `delete`, `index`, or `update`
  action.

* A <<docs-delete,delete API>> request

Instead, you can use the <<docs-update-by-query,update by query>> and
<<docs-delete-by-query,delete by query>> APIs to update or delete existing
documents in a data stream. See <<update-docs-in-a-data-stream-by-query>> and <<delete-docs-in-a-data-stream-by-query>>.

> Alternatively, you can update or delete a document by submitting requests to the

Reviewer: I might say, "If necessary" rather than "Alternatively". It seems like we don't want to encourage folks to directly poke at the backing indices. Their interactions should be almost always be through the data stream, I would think.
